# 1. Context-Sensitive Recommender Systems

----------------------------------------------------------------

_“For me context is the key – from that comes the understanding of everything.”– Kenneth Noland_

----------------------------------------------------------------

**Context-sensitive recommender systems** tailor their recommendations to additional information that defines the specific situation under which recommendations are made. This additional information is referred to as the *context*.

1. **Time:** Recommendations can be affected by many aspects of time, such as weekdays, weekends, holidays, and so on. A recommendation that is relevant to the morning context, may not be relevant in the evening and vice versa.

2. **Location:** With the increasing popularity of GPS-enabled mobile phones, location-sensitive recommendations have gained increasing importance in recent years. For example, a traveling user might wish to determine a recommendation for a restaurant in their locality. Context-sensitive systems can provide more relevant recommendations by using the location as a context.

3. **Social information:** The social context is often important from the perspective of recommender systems. For example, the choice of a user’s friends, tags, and social circles can affect the recommendation process.

In traditional recommender systems with user set $U$ and item set $I$, the set of possibilities in $U × I$ is mapped to a rating. This mapping results in (an incompletely specified) ratings matrix of size $|U|×|I|$. In a context-aware system, an additional set of contextual possibilities is present in a set $C$. For example, the set $C$ might be {morning, afternoon, night}, with the context corresponding to the time of day. In this case, it is no longer possible to map the possibilities in $U × I$ to the ratings, because the same user might have different preferences for an item depending on whether the time is in the morning, afternoon, or night.

Therefore, in context-sensitive recommender systems, the possibilities in $U × I × C$ are mapped to the ratings. Formally, the function $h_R$, which maps the user, items,
and context to the rating, can be written as follows:
$$ h_R : U × I × C → \text{rating} $$

The function $h_{R}$ is subscripted with $R$ to denote the data set to which it is applied. In this case, the ratings data $R$ is a 3-dimensional ratings data cube corresponding to the user, item, and context.

## 1.1. The Multi-dimensional Approach

* **Traditional Recommenders (2D):** <br>
The traditional problem of recommendations can be viewed as that of learning a mapping function from the user-item combinations to the ratings. The corresponding function $f_R$ may be defined as follows:
$$f_R : U × I → \text{rating}$$

* **Generalised Recommenders (Multi-dimensional):** <br>
This general principle motivates the multi-dimensional approach to recommendations, in which the rating problem is seen as that of mapping a set of $w$ different dimensional values to a rating.
$$ g_R : D_1 × D_2 . . . × D_w ⇒ \text{rating} $$

In this case, the ratings data $R$ contain $w$ different dimensions that are mapped to ratings, just as the 2-dimensional user-item combinations are mapped to ratings in the traditional setting. This results in a $w$-dimensional cube rather than a 2-dimensional matrix. The $w$ different dimensions are denoted by $D_1 . . .D_w$.

**Note:** Two of these dimensions will always be users and items and some others might be time, location and so on.

_Therefore, the traditional recommendation problem can be viewed as a special case of the multi-dimensional approach in which the only two dimensions are users and items._

**Example of Contextual Movie Recommender (using Time):**

![Contextual Cube](img/contextual_cube.png)

The rating function $g_R$ is defined as a partial* function, in which the number of arguments is equal to the number of dimensions $w$ (e.g. 3 dimensions = {user, item, time}). In the example above, the rating function $g_R$ (David, Terminator, 9 PM) refers to the rating of user David when he watches the movie Terminator at 9 PM.

*mapping function $g_R$ is referred to as partial because it is defined only for the subset of cells corresponding to observed rating values. The remaining values need to be learned in a data-driven manner for making contextual recommendations.

**Problem Statement:** <br>
 _Determine the top-k possibilities in the “what” (to recommend) dimensions for a particular set of specified values in the “for whom” dimensions._

**Definition (Multi-dimensional Recommendations):** <br>
Given the recommendation space $D_1 × D_2 × . . . D_w$ and the rating function $g_R$: $D_1 × D_2 . . . × D_w → rating$, the recommendation problem is defined by selecting certain “what” dimensions $D_{i1} . . . D_{ip}$ and certain “for whom” dimensions $D_{j1} . . . D_{jq}$ that do not overlap, and recommending for a query tuple $(d_{j1} . . . d_{jq}) ∈ D_{j1} ×. . .× D_{jq} $ the top-k tuples $(d_{i1} . . . d_{ip}) ∈ D_{i1} ×. . .× D_{ip}$ with the maximum predicted value of the rating $g_R(d_1, d_2, . . . , d_w)$.

In other words, a ranked list of the “what” dimension combinations are recommended in response to “for whom” queries.

**Note:** Dimensions such as time can be represented in various granular levels of hierarchy, such as hours, days, weeks, months, and so on. Clearly, the user needs to make careful choices up front about the hierarchy to use, so that the most relevant analysis may be performed in a given application. It is also important to select the most relevant contextual dimensions $D_1 . . . D_w$ for the application at hand. This problem is closely related to that of feature selection in the traditional classification and machine learning literature. Alternatively, these dimensions can be selected by domain experts.

![Contexual Taxonomy](img/contexual_taxonomy.png)

With these hierarchies, one can now make more general (aggregated) queries, such as $g_R\text{(David, Terminator, Evening)}$, instead of $g_R\text{(David, Terminator, 7 PM)}$. The former provides an **average** **prediction** of how much David likes the movie Terminator, if he watches it at any time in the evening, whereas the latter provides a prediction of how much he would like it, if he saw the movie in the 7 PM show.

# 2. Multi-level Multi-dimensional Rating Estimation Problem

Given an initial set of user-assigned ratings specified at different levels of the multi-dimensional cube of ratings, the task is to estimate all other ratings in the cube at all the levels of the OLAP hierarchies.

The techniques for performing contextual recommendation fall into one of three categories:

**1. Contextual pre-filtering:** <br> 
In these methods, a segment of the ratings is pre-filtered corresponding to the relevant context. This relevant segment of ratings is then used to make targeted recommendations.

**2. Contextual post-filtering:** <br>
In these methods, the recommendations are first performed on the entire global set of ratings. Subsequently, the ranked recommendation lists are filtered or adjusted as a post-processing step with the use of temporal context.

**3. Contextual modeling:** <br>
In this case, the contextual information is incorporated directly into the prediction function, rather than as a pre-filtering or post-filtering step.

## 2.1. Contextual **Pre-Filtering** Methods

**Contextual pre-filtering** is also referred to as _reduction_ . In the reduction-based approach, the idea is to reduce the $w$-dimensional estimation problem to a set of 2-dimensional estimations. The 2-dimensional estimation problem is equivalent to that in traditional collaborative filtering systems.

Consider the case where the three attributes are **users (U)**, **movie items (I)**, and **time (T)**. In such a case, the rating function $g_R$ is defined as follows:
$$ g_R : U × I × T → \text{rating} $$

The 3-dimensional ratings estimation can be systematically **reduced to 2-dimensional ratings estimation** on this slice with the following relationship between the 3-dimensional function $g_R$ and the traditional 2-dimensional collaborative filtering function $f_R'(t)$:
$$∀(u, i, t) ∈ U × I × T, g_{R}(u, i, t) = f_{R'(t)}(u, i)$$

The matrix $R'(t)$ is obtained by first selecting the ratings in which the time is fixed to $t$, and then projecting down to the user and item dimensions. In other words, the 2-dimensional slice of the data cube in which the time is fixed to $t$ corresponds to $R'(t)$.

![Contexual Slicing](img/contextual_slicing.png)


Because only a small subset of the ratings are used in a given slice, one may sometimes not have sufficient ratings to perform an accurate recommendation. In such cases, one may aggregate the rating at $t$ with other adjacent time slices to create more accurate recommendations. For example, instead of using $t = 9$ PM, one might use all values of $t$ in the evening, from 7 PM to 11 PM, and then average the ratings in these slices to create the resulting matrix. The 2-dimensional recommender is then applied to this averaged slice.

The **main advantage** of the approach is that it performs the collaborative filtering only over the relevant ratings in which the ratings have been selected with the use of the context. This can lead to improved accuracy in many cases, although the trade-off is that fewer ratings are now being used for prediction.

## 2.2. Contextual **Post-Filtering** Methods

In **post-filtering**, the filtering steps are applied to the output obtained *after* applying a global collaborative filtering algorithm that ignores the contextual information in the data set. In other words, in post-filtering methods, the contextual information is ignored, and a global 2- dimensional ratings matrix is created by aggregating the ratings over all the possible contextual values.

Therefore, the approach comprises two steps:
1. Recommendations are generated on all the data by applying a conventional collaborative filtering model on an _**aggregated**_ user-item matrix. Thus, context is ignored in the first step.

2. Context is then used to **adjust** or **filter** the recommended list.

The multi-dimensional ratings cube aggregates into a 2-dimensional ratings matrix as follows:
* **Explicit ratings:** the aggregation process refers to the averaging of (observed)
ratings.

* **Implicit feedback:** matrices (e.g., units sold), the process of aggregation refers to the sum of the values.

However, when applying traditional collaborative filtering algorithms to this aggregated matrix in order to create predicted ratings $\hat{r}_{uj}$ and a corresponding ranked list of items for each user $u$, the ranked list is not sensitive to the contextual information, because the contextual dimension was ignored in the recommendation process. The post-filtering strategy adjusts the results after the estimations have been made. The adjustments can be made in one of two different ways:

1. Filtering out irrelevant items using heuristic methods.
2. Adjusting the ranking of the recommendations in the list based on the underlying context.

The latter approach can be viewed as a soft version of the former. Both forms of post-filtering **adjust the predicted rating** $\hat{r}_{uj}$ for a given user-item combination.

### 2.2.1. Filtering Method

As mentioned, The idea behind post-filtering is straightforward – we produce the predicted ratings or the list of top-$N$ items without considering contexts by the traditional recommendation algorithms. Then, we can remove the items that are irrelevant to the contexts.

Panniello, et al. (Panniello et al. 2009) proposed the first post-filtering method which can be described as:
$$ 
\hat{r}(u, i, c) = 
\begin{cases} 
\hat{r}(u, i) & \text{if } Pr(u, i, c) \geq p \\ 
0 & \text{if } Pr(u, i, c) < p 
\end{cases}
$$

for user ($u$), item ($i$) and context ($c$).

* The $\hat{r}(u, i, c)$ refers to the predicted rating for the user $u$ on item $i$ within context situation $c$.
* The $\hat{r}(u, i)$ is the predicted rating without considering contexts by a traditional recommendation algorithm. 
* The $Pr(u, i, c)$ is an additional calculated probability, with which the user will choose a certain type of item in a given context.

This probability is computed as the number of neighbors (i.e., users similar to $u$) who purchased or consumed the same item $i$ in contexts $c$ divided by the number of the total number of neighbors. Using movies as an example,

$$
Pr(u, i, c) = \frac{\text{Number of neighbors who watched movie } i \text{ in context } c}{\text{Total number of neighbors}}
$$

* The $p$ is the minimum threshold to be selected. In real-world settings, $p$ is defined with empirical analysis, business knowledge, A/B testing or by optimising for specific (business) metrics. The choice of $p$ is ultimately a balance between relevance (ensuring the user sees items they’re likely to engage with) and exploration (introducing variety). 

We set $\text{r}(u, i, c)$ as zero if this probability is smaller than the threshold $p$, to indicate that the item $i$ is not qualified to be recommended. Otherwise, the model will use $\hat{r}(u, i)$ to represent $\hat{r}(u, i, c)$.

This method is only valid for evaluating the top-$N$ recommendations as it removes irrelevant items associated with the contexts, since they mark  $R(u, i, c)$ as zero if item $i$ is not appropriate to be recommended.

### 2.2.2. Adjusting Method

Ramirez, et al. (Ramirez-Garcia and Garca-Valdez 2014) made the second attempt and they proposed a post-filtering method by adjusting the predicted ratings.

The predicted rating of a user $u$ for item $i$ in context $c$ is defined as follows:
$$
\hat{r}(u, i, c) = \beta \cdot \hat{r}(u, i) + (1 - \beta) \cdot \overline{r}(i, c)
$$

* The $\overline{r}(i, c)$ denotes the average value of the ratings that are placed on item $i$ without context $c$.
* The $\hat{r}(u, i)$ is the predicted rating without considering contexts by a traditional recommendation algorithm (e.g. collaborative filtering). 
* The $\beta$ parameter is a ratio parameter $(0 < \beta < 1)$ to control the contribution of each part.

The parameter $\beta$ can be selected similarly as the parameter $p$ in the filtering method above.

The proposed method above adjusts the predicted ratings compared to the filtering method which filters items.


As a summary, to "adjust" the ratings based on context, a model can be built that predicts the probability $Pr(u, i, c)$ that **user** $u$ will find **item** $i$ relevant within **context** $c$, where this probability represents how well the item fits the context, helping us rank it more appropriately.

$$
\text{Adjusted rating} = Pr(u, i, c) \cdot \hat{r}_{ui}
$$

- This probability can be estimated using various techniques, including:
  - **Content-based models** that focus on item attributes.
  - **Collaborative filtering with pre-filtering** that narrows down recommendations based on context-specific data.
To adjust the ranking based on context, we combine this rating $\hat{r}_{ui}$ with the context-based probability.

Another post filtering method using rating deviations can be found [here](https://cdn.aaai.org/ocs/17661/17661-77652-1-PB.pdf).

**Note:** Post-filtering can be more robust than pre-filtering in a larger number of situations because one combines the local information $Pr(u, i, c)$ with the rating $\hat{r}_{ui}$ determined using all the data.

# 3. Contextual Modeling

In both pre-filtering and post-filtering, the collaborative filtering problem is reduced to the 2-dimensional setting, and the context is used during pre-processing or post-processing. The **main disadvantage** of this approach is that context is not integrated very tightly into the recommendation algorithm. It is possible to incorporate context directly into the recommendation process by modifying existing models (such as neighborhood-based methods) to the $w$-dimensional setting.

## 3.1. Neighborhood-Based Methods

Assume we are interested in predicting the ratings in a 3-dimensional space where the axes represent **users**, **items**, and **context** (e.g., **time**). <br>
A point within this 3D cube could be denoted as $A = (u, i, t)$:

We define the predicted rating as following:
$$
\text{Predicted Rating} = \frac{\sum_{k=1}^{r} \left(\text{Rating of } B_k \cdot \frac{1}{Dist(A, B_k)}\right)}{\sum_{k=1}^{r} \frac{1}{Dist(A, B_k)}}
$$

where $B_k$ is each of the closest neighbors, and $\frac{1}{Dist(A, B_k)}$ is the weight for each neighbor $B_k$.

1. **Calculation of Distances: $Dist(A, B)$** <br>

Consider two points in the 3-dimensional cube, corresponding to $A = (u, i, t)$ and $B = (u', i', t')$ respectively. Then, the distance between $A$ and $B$ can be defined as the sum of the weighted distances (or weighted Euclidean metric alternatively) between the individual dimensions. Using a weighted average rather than a typical Euclidean distance is essential in contextual recommender systems because different dimensions—like user, item, and time—don’t always contribute equally to predicting preferences. Weights allow the model to adapt to the importance of each dimension, which is often context-dependent and can vary across use cases.

* **Weighted Distances:** <br>
$Dist(A,B) = w_1 · Dist(u, u') + w_2 · Dist(i, i') + w_3 · Dist(t, t)$

* **Euclidean weighted Distances:** <br>
$Dist(A,B) = \sqrt{w_1 · Dist(u, u')^2 + w_2 · Dist(i, i')^2 + w_3 · Dist(t, t')^2}$

where:
- $w_1$, $w_2$, and $w_3$ are the weights for the user, item, and time dimensions, respectively.
- $Dist(u, u')$, $Dist(i, i')$, and $Dist(t, t')$ are individual distances for each dimension.


2. **Calculation of Distances: $Dist(u, u')$, $Dist(i, i')$, $Dist(t, t)$** <br>

There are different ways to calculate the distances. The following are some of the techniques that can be used:
- **Collaborative:** In this case, one can use the Pearson method or the adjusted cosine to calculate $Dist(u, u')$, $Dist(i, i')$, $Dist(t, t)$.
- **Content-based:** In this case, the attributes associated with the dimensions (i.e., user profiles and item profiles) are used to compute the profile.
- **Combined:** It is possible to combine the collaborative and content-based measures to obtain a more robust measure of similarity.

**3. Calculation of Weights: $w_1$, $w_2$, $w_3$** <br>

1. **Domain Expertise**: Initial weights can be set based on domain knowledge if a dimension is known to be particularly influential.
2. **Hyperparameter Tuning**: Using techniques like grid search to test different weight combinations on a validation set.
3. **Learning from Data**: Certain models can learn the weights directly as trainable parameters.

### 3.1.1. Example Calculation

Assume:
- $Dist(u, u') = 0.3$,
- $Dist(i, i') = 0.5$,
- $Dist(t, t') = 0.2$,
and weights:
- $w_1 = 0.4$,
- $w_2 = 0.3$,
- $w_3 = 0.3$.

**Weighted Euclidean Distance**
$$
Dist(A, B) = \sqrt{0.4 \cdot 0.3^2 + 0.3 \cdot 0.5^2 + 0.3 \cdot 0.2^2} = \sqrt{0.036 + 0.075 + 0.012} \approx 0.35
$$

**Predicted Rating (using 3 closest neighbors)**
Assume ratings for the 3 closest neighbors $B_1$, $B_2$, and $B_3$ are 4.0, 3.0, and 5.0, with distances:
- $Dist(A, B_1) = 0.5$,
- $Dist(A, B_2) = 0.3$,
- $Dist(A, B_3) = 0.2$.

1. **Compute Weights**:
   - Weight for $B_1 = \frac{1}{0.5} = 2$,
   - Weight for $B_2 = \frac{1}{0.3} \approx 3.33$,
   - Weight for $B_3 = \frac{1}{0.2} = 5$.

2. **Weighted Average**:
   $$
   \text{Predicted Rating} = \frac{4.0 \times 2 + 3.0 \times 3.33 + 5.0 \times 5}{2 + 3.33 + 5} \approx \frac{8 + 9.99 + 25}{10.33} \approx 4.13
   $$

Thus, the predicted rating for $A$ is approximately 4.13.


## 3.2. Factorisation Machines with Contextual Information

In factorization machines, the basic idea is to **model each rating as a linear combination of interactions between input variables**. The input variables are derived from the original ratings matrix.

For example, consider the case in which we have a 3-dimensional cube containing $m$ users, $n$ items, and $d$ values of the contextual dimension, and each rating is associated with a unique triplet. One can then “flatten” this 3-dimensional cube into a set of ($m + n + d$)-dimensional rows, such that each row corresponds to the user, item, and contextual value of an observed rating. Therefore, there are as many rows as the number of observed ratings. We represent the variables of the row by $x_1 . . . x_{m+n+d}$, all of which are either $0s$ or $1s$.

![Contexual Factorization Machines](img/contextual_factorization_machines.png)


**Note:** At first sight, it would seem that we could use a classification or regression predictor on this flattened representation in a straightforward way; however, it would not work very well because of the extraordinary data sparsity, in which there are only three nonzero entries in each row. It is here that factorization machines rescue us from the perils of sparsity.

**Factorisation Machines (FM)** model formulates rating prediction as a regression problem in which user, item, and additional contextual information are combined into a feature vector $\mathbf{x}_i$.  The predictor consists of global bias, first-order, and second-order interactions of the input features.  The estimation is as follow:

$$
\hat{y}(\mathbf{x}_i) = w_0 + \sum_{f=1}^{p} w_f . x_{if} + \sum_{f=1}^{p} \sum_{g=f+1}^{p} x_{if} . x_{ig} \Big( \sum_{k=1}^{K} v_{fk} . v_{gk} \Big)
$$

where $p=(m+n+d)$ is the length of feature vectors and $K$ is the dimensionality of second-order latent factors.

The variables to be learned are:
1. $w_0$ = Global Bias Variable
2. $w_f$ = Feature Covariates
3. $v_i$ = Latent Vectors

Without the additional features and with only user and item, second-order FM is reduced to matrix factorization.  Therefore, FM is a general framework to incorporate contextual information while maintaining the effectiveness of factorization models in recommender systems.

**Note:** FM can be extended and generalized into higher-order interactions.  Given the degree of data sparsity commonly faced in recommender systems, second-order FM is usually sufficient. Higher orders would be less efficient and harder to estimate.


To learn the parameters of the FM regressor, we minimize the following regularized squared loss function using Stohastic Gradient Descent (SGD):
$$
\mathcal{L}(\mathbf{w, V} | \lambda) = \frac{1}{2} \sum_{(\mathbf{x}, y) \in \mathcal{D}} (y - \hat{y}(\mathbf{x})) + \frac{\lambda}{2} || \mathbf{w} ||^2 + \frac{\lambda}{2} ||\mathbf{V}||_2^2
$$

# 4. Python Implementation

In [1]:
# Import common python libraries
import pandas as pd
from sklearn.model_selection import train_test_split

# Cornac is a python library to implement non-contextual recommenders (used for comparison)
import cornac
from cornac.utils import cache

print(f"Cornac version: {cornac.__version__}")

/Users/atawua/Documents/03-DATA-SCIENCE/01-github/movielens-100k-recommender-systems/.venv/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Cornac version: 2.2.2


In [16]:
# Install libfm library needed to train a contextual awareness model
!git clone https://github.com/srendle/libfm.git
!make all -C libfm

Cloning into 'libfm'...
remote: Enumerating objects: 233, done.
remote: Total 233 (delta 0), reused 0 (delta 0), pack-reused 233 (from 1)
Receiving objects: 100% (233/233), 129.46 KiB | 940.00 KiB/s, done.
Resolving deltas: 100% (112/112), done.
cd src/libfm; make all
g++ -O3 -Wall -c libfm.cpp -o libfm.o
In file included from libfm.cpp:57:
In file included from ./src/fm_learn_mcmc_simultaneous.h:43:
./src/fm_learn_mcmc.h:739:8: warning: variable 'num_all' set but not used [-Wunused-but-set-variable]
  739 |   uint num_all = 0;
      |        ^
./src/fm_learn_mcmc.h:854:8: warning: variable 'num_all' set but not used [-Wunused-but-set-variable]
  854 |   uint num_all = 0;
      |        ^
2 warnings generated.
mkdir -p ../../bin/
g++ -O3 -Wall libfm.o -o ../../bin/libFM
g++ -O3 -Wall -c tools/transpose.cpp -o tools/transpose.o
mkdir -p ../../bin/
g++ -O3 tools/transpose.o -o ../../bin/transpose
g++ -O3 -Wall -c tools/convert.cpp -o tools/convert.o
mkdir -p ../../bin/
g++ -O3 tools/conv

In [2]:
# Cache the data
cache("http://files.grouplens.org/datasets/hetrec2011/hetrec2011-movielens-2k-v2.zip", unzip=True)

Data from http://files.grouplens.org/datasets/hetrec2011/hetrec2011-movielens-2k-v2.zip
will be cached into /Users/atawua/.cornac/hetrec2011-movielens-2k-v2.zip


18.9MB [00:01, 10.1MB/s]                            


Unzipping ...
File cached!


'/Users/atawua/.cornac/hetrec2011-movielens-2k-v2.zip'

In [5]:
# Generate the data directory path
DATA_DIRECTORY =  "/Users/atawua/.cornac"
DATA_FILE = "/user_ratedmovies.dat"

DATA_PATH = DATA_DIRECTORY + DATA_FILE
DATA_PATH

'/Users/atawua/.cornac/user_ratedmovies.dat'

In [6]:
# Load the data and get a quick view of it
user_ratedmovies_df = pd.read_csv(DATA_PATH, sep="\t")
user_ratedmovies_df.head()

userID  movieID  rating  date_day  date_month  date_year  date_hour  \
0      75        3     1.0        29          10       2006         23   
1      75       32     4.5        29          10       2006         23   
2      75      110     4.0        29          10       2006         23   
3      75      160     2.0        29          10       2006         23   
4      75      163     4.0        29          10       2006         23   

   date_minute  date_second  
0           17           16  
1           23           44  
2           30            8  
3           16           52  
4           29           30

In [7]:
# Show data statistics
print("Number of users:", user_ratedmovies_df["userID"].nunique())
print("Number of movies:", user_ratedmovies_df["movieID"].nunique())

(
user_ratedmovies_df
    .groupby('userID')
    .agg({'movieID': 'count'})
    .rename({"movieID": "total_number_of_rated_movies"}, axis=1)
    .sort_values(by='total_number_of_rated_movies', ascending=False)
    )

Number of users: 2113
Number of movies: 10109


total_number_of_rated_movies
userID                              
6757                            3410
30687                           2908
30500                           2823
62332                           2763
51033                           2631
...                              ...
54250                             20
48803                             20
26873                             20
28180                             20
59995                             20

[2113 rows x 1 columns]

In [8]:
# Take a sample of the data (300K) for quicker runs
df = user_ratedmovies_df[["userID", "movieID", "date_hour", "rating"]].sample(n=300_000).copy()
df.head()

userID  movieID  date_hour  rating
17084     1860     2893         16     3.0
158426   13283     1234          2     3.5
555826   44287     2273          1     3.0
608059   49879     1226          8     3.0
639291   51954      597         12     2.5

In [9]:
# Split into train and test sets for evaluation
train_df, test_df = train_test_split(df, test_size=0.2, random_state=42)
print("Training size:", len(train_df))
print("Test size:", len(test_df))

Training size: 240000
Test size: 60000


## 4.1. Factorized Machines with Context (time)

In [10]:
def convert_df_to_flatten(df, columns=["userID", "movieID", "date_hour"]):
    return pd.get_dummies(df, columns=columns).astype(int)

# Apply one-hot encoding to 'userID', 'movieID', and 'date_hour'
train_sparse_df = convert_df_to_flatten(train_df)
test_sparse_df = convert_df_to_flatten(test_df)

# Display the resulting sparse DataFrame
train_sparse_df.head()

rating  userID_75  userID_78  userID_127  userID_170  userID_175  \
266989       3          0          0           0           0           0   
158068       4          0          0           0           0           0   
552454       2          0          0           0           0           0   
658846       3          0          0           0           0           0   
432255       1          0          0           0           0           0   

        userID_190  userID_267  userID_325  userID_383  ...  date_hour_14  \
266989           0           0           0           0  ...             0   
158068           0           0           0           0  ...             0   
552454           0           0           0           0  ...             0   
658846           0           0           0           0  ...             0   
432255           0           0           0           0  ...             0   

        date_hour_15  date_hour_16  date_hour_17  date_hour_18  date_hour_19  \
266989             0             0             0             0             0   
158068             0             0             0             0             1   
552454             0             0             0             0             0   
658846             0             0             0             0             0   
432255             0             1             0             0             0   

        date_hour_20  date_hour_21  date_hour_22  date_hour_23  
266989             0             0             0             0  
158068             0             0             0             0  
552454             0             0             0             0  
658846             0             0             0             0  
432255             0             0             0             0  

[5 rows x 11111 columns]

In [11]:
# Confirm that flatten is successful (columns = users + movies + context + target)
assert train_sparse_df.shape[1] == train_df["userID"].nunique() + train_df["movieID"].nunique() + train_df["date_hour"].nunique() + 1

The first thing that we need to do is prepare the data to be consumed by the [LibFM](http://www.libfm.org/libfm-1.42.manual.pdf) python library. The data format being used by Libfm expects the following format:

![Libfm_format](img/libfm_text_format.png)

The first column states the target of each of the three case: i.e. rating=4 for the first case, rating=2 for the second and rating=-1 for the third. After the target, each line contains the non-zero elements of $x$, where an entry like 0:1.5 reads $x_0 = 1.5$ and 3:-7.9 means $x_3 = −7.9$, etc. That means the left side of INDEX:VALUE states the index within $x$ whereas the right side states the value of $x_\text{INDEX}$, i.e. $x_\text{INDEX}$ = VALUE.

The above example would equal to the following matrix representations:
$$
X = \begin{pmatrix}
1.5 & 0.0 & 0.0 & -7.9 & 0.0 & 0.0 & 0.0 \\
0.0 & 10^{-5} & 0.0 & 2.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 1.0 \\
\end{pmatrix},
\quad
y = \begin{pmatrix}
4 \\
2 \\
-1 \\
\end{pmatrix}
$$


In [12]:
def convert_df_to_libfm_format(df):
    """ Converts a pandas data frame to the expected data format of the libFM library """
    
    # Initialize an empty list to store each row in the required format
    fm_format_data = []
    
    # Iterate through each row in the DataFrame
    for _, row in df.iterrows():
        # Start the line with the rating as the target value
        line = [str(row['rating'])]
        
        # Iterate over each feature column (except 'rating') and find non-zero entries
        for j, (col, value) in enumerate(row.items()):
            if col != 'rating' and value != 0:  # Only include non-zero entries
                line.append(f"{j-1}:{value}")   # Adjusting index by -1 to account for "rating" column
        
        # Join the line elements with spaces and add to the output list
        fm_format_data.append(" ".join(line))
    
    return fm_format_data

In [13]:
# Generate the libFM formatted data
fm_train_df = convert_df_to_libfm_format(train_sparse_df)
fm_test_df = convert_df_to_libfm_format(test_sparse_df)

# Print the first few lines to verify
for line in fm_train_df[:5]:
    print(line)
    
# NOTE: Takes about ~10min to run

3 681:1 7756:1 11098:1
4 392:1 3572:1 11105:1
2 1312:1 4043:1 11091:1
3 1574:1 4127:1 11087:1
1 1019:1 5444:1 11102:1


In [ ]:
# # Uncomment if interested to perform sanity checks based on above output and libFM data expectations
# train_sparse_df.drop('rating', axis=1).iloc[i, j]

In [14]:
# Save train data as train.libfm
with open("train.libfm", "w") as train_file:
    for line in fm_train_df:
        train_file.write(line + "\n")

# Save test data as test.libfm
with open("test.libfm", "w") as test_file:
    for line in fm_test_df:
        test_file.write(line + "\n")

In [17]:
# Explore usage of the libFM parameters
!./libfm/bin/libFM

----------------------------------------------------------------------------
libFM
  Version: 1.4.4
  Author:  Steffen Rendle, srendle@libfm.org
  WWW:     http://www.libfm.org/
This program comes with ABSOLUTELY NO WARRANTY; for details see license.txt.
This is free software, and you are welcome to redistribute it under certain
conditions; for details see license.txt.
----------------------------------------------------------------------------
-cache_size     cache size for data storage (only applicable if data is
                in binary format), default=infty
-dim            'k0,k1,k2': k0=use bias, k1=use 1-way interactions,
                k2=dim of 2-way interactions; default=1,1,8
-help           this screen
-init_stdev     stdev for initialization of 2-way factors; default=0.1
-iter           number of iterations; default=100
-learn_rate     learn_rate for SGD; default=0.1
-load_model     filename for reading the FM model
-meta           filename for meta information about dat

The most important parameter here is the **"dim" parameter** representing $k_0$, $k_1$ and $k_2$ which controls the dimensionality of the factorization machine model.

Let's recall our Loss function and how the $k$ parameters are connected:
$$
\hat{y}(\mathbf{x}_i) = w_0 + \sum_{f=1}^{p} w_f \cdot x_{if} + \sum_{f=1}^{p} \sum_{g=f+1}^{p} x_{if} \cdot x_{ig} \left( \sum_{k=1}^{K} v_{fk} \cdot v_{gk} \right)
$$

1. **$w_0$** is the global bias term, controlled by $k_0$.
   - When $k_0 = 1$, $w_0$ is included. This parameter represents an overall bias for all predictions (the baseline level).
   - When $k_0 = 0$, the bias term is excluded from the model.

2. **$\sum_{f=1}^{p} w_f \cdot x_{if}$** represents one-way (linear) interactions for each feature, controlled by $k_1$.
   - If $k_1 = 1$, each feature $x_{if}$ has a weight $w_f$, and their linear contributions are added to the prediction.
   - If $k_1 = 0$, these linear terms are excluded, and the model focuses only on interactions between features.

3. **$\sum_{f=1}^{p} \sum_{g=f+1}^{p} x_{if} \cdot x_{ig} \left( \sum_{k=1}^{K} v_{fk} \cdot v_{gk} \right)$** represents two-way (pairwise) interactions between features, controlled by $k_2$, where:
   - $v_{fk}$ and $v_{gk}$ are latent factors that capture hidden relationships between features.
   - $k_2$ (the dimensionality of the latent factors) controls how complex or nuanced these interactions can be.

The choice of **$k_2$** depends on the complexity of your data and the capacity of the model you want to build:

- Small values of $k_2$ (e.g., 2-10):
  - Less complex interactions: Good for smaller or simpler datasets.
  - Less computationally expensive: Reduces memory and computational requirements, making the model faster to train.

- Larger values of $k_2$ (e.g., 20-100):
  - More complex interactions: Allows the model to capture more detailed and nuanced interactions between features.
  - More computationally intensive: Requires more memory and may take longer to train.

In [18]:
# Train the Contextual Awareness recommender system and validating on the test set.
!./libfm/bin/libFM -task r -train train.libfm -test test.libfm -seed 42 -dim "1,1,10" -iter 200

# Parameters:
# 1. -dim: 1 (Train global bias), 1 (Use 1-way interactions), 10 (Use 10-latent factor columns)
# 2. -iter: 200 (Perform 200 iterations for training)

----------------------------------------------------------------------------
libFM
  Version: 1.4.4
  Author:  Steffen Rendle, srendle@libfm.org
  WWW:     http://www.libfm.org/
This program comes with ABSOLUTELY NO WARRANTY; for details see license.txt.
This is free software, and you are welcome to redistribute it under certain
conditions; for details see license.txt.
----------------------------------------------------------------------------
Loading train...	
has x = 0
has xt = 1
num_rows=240000	num_values=720000	num_features=11110	min_target=0	max_target=5
Loading test... 	
has x = 0
has xt = 1
num_rows=60000	num_values=180000	num_features=8906	min_target=0	max_target=5
#relations: 0
Loading meta data...	
#Iter=  0	Train=1.01875	Test=1.06367
#Iter=  1	Train=0.913931	Test=1.05874
#Iter=  2	Train=0.876879	Test=1.06844
#Iter=  3	Train=0.858092	Test=1.08006
#Iter=  4	Train=0.847626	Test=1.09513
#Iter=  5	Train=0.840836	Test=1.10753
#Iter=  6	Train=0.837865	Test=1.11934
#Iter=  7	Train=

We can see the the **RMSE** for the test set is: **1.2256**

## 4.2. Factorised Machines without Context

In [19]:
VERBOSE = True
SEED = 42

# 1. Define the Matrix Factorisation model
mf = cornac.models.MF(
    k=10, 
    max_iter=20, 
    learning_rate=0.01, 
    lambda_reg=0.02, 
    use_bias=True,
    verbose=VERBOSE, 
    seed=SEED,
)

# 2. Define the evaluation method of the model (i.e. train-test split)
eval_method = cornac.eval_methods.BaseMethod.from_splits(
    train_data=list(train_df.itertuples(index=False)), 
    test_data=list(test_df.itertuples(index=False)),
    exclude_unknowns=False, 
    verbose=VERBOSE,
    seed=SEED,
)

rating_threshold = 1.0
exclude_unknowns = False
---
Training data:
Number of users = 2113
Number of items = 8973
Number of ratings = 240000
Max rating = 23.0
Min rating = 0.0
Global mean = 12.1
---
Test data:
Number of users = 2113
Number of items = 9220
Number of ratings = 60000
Number of unknown users = 0
Number of unknown items = 247
---
Total users = 2113
Total items = 9220


In [20]:
# Train and evaluate the RMSE of the Matrix factorisation model
test_result, _ = eval_method.evaluate(
    model=mf, 
    metrics=[cornac.metrics.RMSE()],
    user_based=False,
)
print(test_result)


[MF] Training started!


100%|██████████| 20/20 [00:00<00:00, 287.64it/s, loss=3348350.25]


Optimization finished!

[MF] Evaluation started!


Rating: 100%|██████████| 60000/60000 [00:00<00:00, 208216.91it/s]

   |   RMSE | Train (s) | Test (s)
-- + ------ + --------- + --------
MF | 7.3402 |    0.4122 |   0.2982



We can see the the **RMSE** for the test set is: **7.3402**

**Conclusion:** We can see that the **Contextual Model (RMSE = 1.2256)** *outperformed* the **Non-Contextual Model (RMSE = 7.3402)**. <br>
In other words, it provided more **accurate** and **relevant** predictions of each user.

# 5. Summary

There are three primary ways in which context-aware recommendations are performed.

* **Pre-filtering:** 
The problem is reduced to a 2-dimensional collaborative filtering problem by filtering the w-dimensional cube to a 2-dimensional ratings matrix before applying the collaborative filtering algorithm.

* **Post-filtering:** 
The context is ignored during the first phase of collaborative filtering. Subsequently, the results are adjusted with the use of a predictive model that regulates the relative importance of context. 

* **Contextual Modelling:**
Finally, a recent approach is to incorporate the context directly into the model by treating it as a $w$-dimensional prediction problem. Generalizations of matrix factorization and linear regression models have been proposed in this setting. This approach is computationally intensive, but it is a generic approach with the best potential when a large amount of data is available.